pex/fetcher.py

```diff
@@ -0,0 +1,90 @@
+# Copyright 2021 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import absolute_import
+
+import ssl
+import time
+from contextlib import closing, contextmanager
+
+from pex.compatibility import FileHandler, HTTPError, HTTPSHandler, ProxyHandler, build_opener
+from pex.network_configuration import NetworkConfiguration
+from pex.typing import TYPE_CHECKING, cast
+
+if TYPE_CHECKING:
+    from typing import BinaryIO, Dict, Iterator, Optional, Text
+else:
+    BinaryIO = None
+
+
+class URLFetcher(object):
+    def __init__(
+        self,
+        network_configuration=None,  # type: Optional[NetworkConfiguration]
+        handle_file_urls=False,  # type: bool
+    ):
+        # type: (...) -> None
+        network_configuration = network_configuration or NetworkConfiguration()
+
+        self._timeout = network_configuration.timeout
+        self._max_retries = network_configuration.retries
+
+        ssl_context = ssl.create_default_context(cafile=network_configuration.cert)
+        if network_configuration.client_cert:
+            ssl_context.load_cert_chain(network_configuration.client_cert)
+
+        proxies = None  # type: Optional[Dict[str, str]]
+        if network_configuration.proxy:
+            proxies = {protocol: network_configuration.proxy for protocol in ("http", "https")}
+
+        handlers = [ProxyHandler(proxies), HTTPSHandler(context=ssl_context)]
+        if handle_file_urls:
```

Comment:
The only changes to this class besides being moved from requirements.py to here are:
1. Drop "ftp" from the proxy protocols above since we don't install the FtpHandler / never intended to support plaintext FTP.
2. Add optional FileHandler support which will be needed / useful for handling --find-links "downloads" from local find-links repos in a way that keeps manual dist hashing code uniform / simpler.